### PR TITLE
Modifying Replay Memory and Prioritized Replay Memory

### PR DESCRIPTION
**Pull request by mclearning2**
modify Replay buffer and Prioritized replay buffer

I expect `np.ndarray` to be faster than `list`. This is because while `np.ndarray` copies original data,  `list` refers to the data. In `add()` function `np.ndarray` will be slower than `list` due to hard copy operation. However, in `sample()` function, each component of `list` is located nearby but it's not real data. it's address of the data. So it takes more time to refer to the data than `nd.array`.

The `NStepTransitionBuffer` gets and stores demo data. It doesn't match function with name. I think it'll be better to implement one `ReplayBuffer` which has 2 function. Also I applied `PrioritizedReplayBuffers` for the same reason.

I compared all the agents to see if the time cost is getting smaller. The result, however, is that the smaller the buffer, the smaller the difference of the time cost. So Even a small gap, my code is still better than before, because I've simplified the code by combining function of demonstration, n-step.

### CODE REVIEW

**Comment by mclearning2** [comment]
please check and help me out!

File: algorithms/common/buffer/priortized_replay_buffer.py [hunk 1]
```diff
@@ -59,27 +71,36 @@ def __init__(self, buffer_size: int, batch_size: int, alpha: float = 0.6):
...
+
     def add(
         self,
-        state: np.ndarray,
-        action: np.ndarray,
-        reward: np.float64,
-        next_state: np.ndarray,
-        done: bool,
-    ):
+        transition: Tuple[np.ndarray, Union[int, np.ndarray], float, np.ndarray, bool],
```
Comment (MrSyee) [inline algorithms/common/buffer/priortized_replay_buffer.py]: The type of action would be one thing rather than Union.

Comment (mclearning2) [reply algorithms/common/buffer/priortized_replay_buffer.py]: The types of action vary depending on the environment. For example, PongDeterministic-v4 has action of int type. However, In the Pendulum-v0, the action type is np.ndarray of float32.

File: algorithms/common/buffer/replay_buffer.py [hunk 2]
```diff
@@ -15,169 +15,149 @@
...
-        action: np.ndarray,
-        reward: np.float64,
-        next_state: np.ndarray,
-        done: float,
+        buffer_size: int,
+        batch_size: int = 32,
+        gamma: float = 0.99,
+        n_step: int = 1,
+        demo: List[
+            Tuple[np.ndarray, Union[int, np.ndarray], float, np.ndarray, bool]
```
Comment (MrSyee) [inline algorithms/common/buffer/replay_buffer.py]: The type of action in the demo would be one type.

File: algorithms/common/buffer/priortized_replay_buffer.py [hunk 3]
```diff
@@ -26,16 +26,26 @@ class PrioritizedReplayBuffer(ReplayBuffer):
...
+    def __init__(
+        self,
+        buffer_size: int,
+        batch_size: int = 32,
+        gamma: float = 0.99,
+        n_step: int = 1,
+        alpha: float = 0.6,
+        epsilon_d: float = 1.0,
+        demo: List[
+            Tuple[np.ndarray, Union[int, np.ndarray], float, np.ndarray, bool]
```
Comment (MrSyee) [inline algorithms/common/buffer/priortized_replay_buffer.py]: The type of action in the demo would be one type.

File: algorithms/common/buffer/priortized_replay_buffer.py (hunk 1 above)
Comment (MrSyee) [reply algorithms/common/buffer/priortized_replay_buffer.py]: I mean when data is added buffer, already changes the type on agent class.

File: algorithms/common/buffer/replay_buffer.py [hunk 4]
```diff
@@ -15,169 +15,149 @@
...
-    """
-
-    def __init__(self, buffer_size: int, n_step: int, gamma: float, demo: list = None):
         """Initialize a ReplayBuffer object.
 
         Args:
             buffer_size (int): size of replay buffer for experience
-            demo (list): demonstration transitions
-
+            batch_size (int): size of a batched sampled from replay buffer for training
```
Comment (MrSyee) [inline algorithms/common/buffer/replay_buffer.py]: should write the other args.

**Comment by Curt-Park** [comment]
Please consider to add wandb config update for args.

File: algorithms/common/buffer/priortized_replay_buffer.py [hunk 5]
```diff
@@ -26,16 +26,24 @@ class PrioritizedReplayBuffer(ReplayBuffer):
     https://github.com/openai/baselines/blob/master/baselines/deepq/replay_buffer.py
```
Comment (Curt-Park) [inline algorithms/common/buffer/priortized_replay_buffer.py]: need to modify this comment.
We changed a lot of the implementation. (not taken)

**Comment by Curt-Park** [comment]
Please verify the modification on the proper environments and attach the logs with the baseline commit id.

**Comment by Curt-Park** [comment]
Please fill out the PR body. See the number 7 of this reference: https://chris.beams.io/posts/git-commit/

File: algorithms/common/buffer/priortized_replay_buffer.py (hunk 1 above)
Comment (mclearning2) [reply algorithms/common/buffer/priortized_replay_buffer.py]: I fixed it. But demo's action is not yet np.ndarray type. We have to make or get new demo data.

File: algorithms/common/buffer/priortized_replay_buffer.py [hunk 6]
```diff
@@ -59,27 +69,33 @@ def __init__(self, buffer_size: int, batch_size: int, alpha: float = 0.6):
...
     def add(
-        self,
-        state: np.ndarray,
-        action: np.ndarray,
-        reward: np.float64,
-        next_state: np.ndarray,
-        done: bool,
-    ):
+        self, transition: Tuple[np.ndarray, np.ndarray, float, np.ndarray, bool]
+    ) -> Union[Tuple[np.ndarray, np.ndarray, float, np.ndarray, bool], Tuple[()]]:
```
Comment (MrSyee) [inline algorithms/common/buffer/priortized_replay_buffer.py]: how about just tuple?

File: algorithms/common/buffer/replay_buffer.py [hunk 7]
```diff
@@ -15,169 +15,152 @@
...
+                self.acts_buf[idx] = action
+                self.rews_buf[idx] = reward
+                self.next_obs_buf[idx] = next_state
+                self.done_buf[idx] = done
 
-    def add(self, transition: Tuple[np.ndarray, ...]) -> Tuple[Any, ...]:
-        """Add a new transition to memory."""
+    def add(
+        self, transition: Tuple[np.ndarray, np.ndarray, float, np.ndarray, bool]
+    ) -> Union[Tuple[np.ndarray, np.ndarray, float, np.ndarray, bool], Tuple[()]]:
```
Comment (MrSyee) [inline algorithms/common/buffer/replay_buffer.py]: I think Tuple[Any, ...] is better.

**Comment by Curt-Park** [comment]
Please attach the results of other methods that use the buffer.

**Comment by mclearning2** [comment]
It's similar in DQfD

### [DQfD lunarlander_v2](https://app.wandb.ai/medipixel_rl/lunarlander_v2?workspace=user-mclearning2): buffer size = 100,000

W B Chart 8_26_2019, 8_39_04 AM
W B Chart 8_26_2019, 8_40_37 AM

**Comment by mclearning2** [comment]
### [DQfD lunarlander_v2](https://app.wandb.ai/medipixel_rl/lunarlander_v2?workspace=user-mclearning2): buffer size = 100,000

image

**Comment by mclearning2** [comment]
### [bc-DDPG Lunarlander-continuous-v2](https://app.wandb.ai/medipixel_rl/lunarlander_continuous_v2?workspace=user-mclearning2)

W B Chart 8_28_2019, 5_33_29 PM
W B Chart 8_28_2019, 5_33_49 PM

**Comment by mclearning2** [comment]
### [bc-SAC Lunarlander-continuous-v2](https://app.wandb.ai/medipixel_rl/lunarlander_continuous_v2?workspace=user-mclearning2)

W B Chart 8_28_2019, 5_35_52 PM
W B Chart 8_28_2019, 5_35_54 PM

**Comment by mclearning2** [comment]
## Pongnoframeskip-v4

### Rainbow-IQN
W B Chart 9_2_2019, 8_53_29 AM
W B Chart 9_2_2019, 8_53_32 AM

**Comment by Curt-Park** [comment]
@mclearning2 please add smoothing option on the graph or use average values of multiple training results. It’s not so clear to see.

**Comment by mclearning2** [comment]
@Curt-Park Okay I'll smooth all training graph about 0.5. In the result, all score and processing time are very similar to List buffer because of the small buffer size. It may not happen to use a larger buffer size of memory. But my pull request still has some benefits. First, when training starts and the memory is not enough to accommodate transitions, the process is immediately stopped by Numpy `MemoryError`, but List doesn't. Second, ReplayBuffer can handle n-step memory and demonstration easily. So if the results of Reacher environment is good, I think would be great to merge that.

**Comment by Curt-Park** [comment]
@mclearning2
Good job. What was the reason of performance degradation of SAC? Was it just because of seed?

**Comment by mclearning2** [comment]
@Curt-Park I guess so. But in just case, I'll run the sac agent in list version by 5 seeds. It would be not taken long.

**Comment by Curt-Park** [comment]
@mclearning Okay. I will give you the review score after checking the result.

File: algorithms/common/buffer/replay_buffer.py [hunk 8]
```diff
@@ -15,169 +15,152 @@
...
-    def __init__(self, buffer_size: int, n_step: int, gamma: float, demo: list = None):
         """Initialize a ReplayBuffer object.
 
         Args:
             buffer_size (int): size of replay buffer for experience
-            demo (list): demonstration transitions
-
+            batch_size (int): size of a batched sampled from replay buffer for training
+            gamma (float): discount factor
+            n_step (int): n_step (int): step size for n-step transition
```
Comment (MrSyee) [inline algorithms/common/buffer/replay_buffer.py]: it is duplicated

**Comment by mclearning2** [comment]
## Reacher-V2

### TD3

W B Chart 9_10_2019, 9_29_01 PM
W B Chart 9_10_2019, 9_29_03 PM

### DDPG

W B Chart 9_10_2019, 9_34_46 PM
W B Chart 9_10_2019, 9_34_48 PM

### SAC

W B Chart 9_10_2019, 9_35_35 PM
W B Chart 9_10_2019, 9_35_37 PM

SAC is running

**Comment by mclearning2** [comment]
## [SAC only list](https://app.wandb.ai/medipixel_rl/lunarlander_continuous_v2?workspace=user-mclearning2)
I ran the SAC agent five times without interlim-test. As the result, It's different by seed. `sac(list)` is the reult that I ran the last time that to compare with numpy.

### list versions
W B Chart 9_11_2019, 8_33_43 AM
W B Chart 9_11_2019, 8_33_47 AM

### numpy versions
W B Chart 9_11_2019, 8_39_21 AM
W B Chart 9_11_2019, 8_39_23 AM

**Comment by MrSyee** [comment]
Great. I approve of this PR.

**Comment by MrSyee** [comment]
@all-contributors please add @mclearning2  for code

**Pull request merged**